### ISSUE TITLE
ejabberdctl join_cluster keeps failing until I restart the node it tries to join

### ISSUE THREAD

**Issue by joudinet** [processone/ejabberd#3843]
## Environment

- ejabberd version: 21.12
- Erlang version: Erlang (SMP,ASYNC_THREADS) (BEAM) emulator version 10.7.2.14
- OS: Linux (Buildroot)
- Installed from: Buildroot

## Errors from error.log/crash.log

```
2022-06-10 16:59:53.271322+02:00 [error] <0.988.0> Error in process <0.988.0> on node 'ejabberd@testbed-5.gnyoi.com' with exit value:
{shutdown,[{mnesia_loader,handle_exit,2,
                          [{file,"mnesia_loader.erl"},{line,963}]},
           {mnesia_loader,tab_receiver,5,
                          [{file,"mnesia_loader.erl"},{line,440}]},
           {mnesia_loader,spawned_receiver,8,
                          [{file,"mnesia_loader.erl"},{line,343}]}]}

2022-06-10 16:59:53.305192+02:00 [error] <0.985.0> Error in process <0.985.0> on node 'ejabberd@testbed-5.gnyoi.com' with exit value:
{badarg,[{erlang,unlink,[undefined],[]},
         {mnesia_loader,init_receiver,5,
                        [{file,"mnesia_loader.erl"},{line,299}]},
         {mnesia_loader,do_get_network_copy,5,
                        [{file,"mnesia_loader.erl"},{line,221}]},
         {mnesia_controller,'-load_table_fun/1-fun-4-',5,
                            [{file,"mnesia_controller.erl"},{line,2193}]},
         {mnesia_controller,'-load_and_reply/2-fun-0-',2,
                            [{file,"mnesia_controller.erl"},{line,2140}]}]}

```

## Bug description

I have a cluster of 6 ejabberd nodes. At the beginning of the experiment, there are split into 2 clusters of 3 nodes on separated network. When I merge the two networks and use ejabberdctl join_cluster command on the three last nodes of the second cluster to join the first cluster, it often fails (returns a non-zero value).
When it fails, it continues to fail until I restart the remote node I'm trying to join.

For example, I have eja-1, eja-2, and eja-3 on the first cluster:
eja-1$ ejabberdctl join_cluster eja-3.example.com 
eja-2$ ejabberdctl join_cluster eja-1.example.com 

Then, I have eja-4, eja-5, and eja-6 on a second cluster:
eja-4$  ejabberdctl join_cluster eja-5.example.com
eja-6$  ejabberdctl join_cluster eja-4.example.com

So far, so good. Next I merge the two networks and start to move node from the second cluster to the fist one:
eja-4$ ejabberdctl join_cluster eja-1.example.com
eja-5$ ejabberdctl  join_cluster eja-1.example.com # returns non-zero
eja-6$ ejabberdctl  join_cluster eja-1.example.com # returns non-zero

It works for eja-4 but fails for both eja-5 and eja-6. The only difference I see is that both eja-5 and eja-6 detected a partitioned network and restarted while eja-4 was running the join_cluster command.

I connected to eja-1 and asked for an ejabberdctl status and everythting looks good and the list_cluster command does report being in a cluster with eja-1, eja-2, eja-3, and eja-4.
If I try again to join_cluster eja-1 from eja-5 and eja-6, they keep failing until I restart eja-1 node. Any idea for the reason the join_cluster command may keep failing? 

Could there be a race condition between restarting due to a partitioned network and using the join_cluster command?

**Comment by badlop** [comment]
I have very small experience with ejabberd clustering, but reading you, i have a wild speculation:

At mid experiment, Node 4 is a master, being 5 and 6 its slaves, right? When you move away node 4 from that cluster, 5 and 6 complain about partitioned network.

You could try: first move slaves 5 and 6 from one cluster to the other, later move the remaining master node 4.

Maybe it's also worth playing with [leave_cluster](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#leave-cluster) before joining another cluster. Or maybe [set_master](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#set-master).

**Comment by joudinet** [comment]
Thanks @badlop for your insights. I have no idea how to know which node is a master and which nodes are slaves. Do you think it is related with the order I call the join_cluster commands?
So far, I considered there were all equals and I ignored the leave_cluster because I thought it would not be necessary if I'm joining another cluster. I'll try to play with the leave_cluster and set_master commands. If you have any idea how I could known which node is a master, it would help me. Thanks.

**Comment by badlop** [comment]
I had to check ejabberd and erlang documentation. Starting point: https://docs.ejabberd.im/admin/guide/clustering/#clustering-setup

Erlang distribution only cares about nodes connecting, erlang node name and cookie file. The concepts of master/slave appears when reading about mnesia distribution. In fact, those concepts apparently are only relevant here: https://www.erlang.org/doc/man/mnesia.html#set_master_nodes-2

I'll run this in the first node to get information about its mnesia distribution:
```erlang
{mnesia:system_info(db_nodes), mnesia:table_info(passwd, master_nodes), mnesia:system_info(master_node_tables)}.
```

And now for the testing, I setup and start first node called `ejabberd@localhost`, and a second node called `ejabberd2@localhost`. 

On the first node try the functions mentioned earlier in its shell:
```erlang
{[ejabberd@localhost],[],[]}
```

Then on second node run `ejabberdctl join_cluster ejabberd@localhost`. First and second nodes show the same:
```erlang
{[ejabberd2@localhost,ejabberd@localhost],[],[]}
```

Then on second node run `ejabberdctl set_master ejabberd@localhost`. First node still shows:
```erlang
{[ejabberd2@localhost,ejabberd@localhost],[],[]}
```
Second node shows:
```erlang
{[ejabberd2@localhost,ejabberd@localhost],
 [ejabberd@localhost],
 [mqtt_session,push_session,last_activity,pubsub_index,
  vcard_search,oauth_client,archive_prefs,motd_users,mqtt_sub,
  reg_users_counter,mqtt_pub,sr_user,schema,roster_version,
  pubsub_orphan,session,pubsub_last_item,offline_msg,
  archive_msg,route,private_storage,motd,sr_group,oauth_token,
  bytestream,muc_room|...]}
```

Now, after running in first node `ejabberdctl set_master ejabberd2@localhost`, this first node shows:
```erlang
{[ejabberd2@localhost,ejabberd@localhost],
 [ejabberd2@localhost],
 [mqtt_session,push_session,last_activity,pubsub_index,
  vcard_search,oauth_client,archive_prefs,motd_users,mqtt_pub,
  reg_users_counter,mqtt_sub,sr_user,schema,roster_version,
  pubsub_orphan,session,pubsub_last_item,offline_msg,
  archive_msg,route,private_storage,motd,sr_group,oauth_token,
  bytestream,muc_room|...]}
```
and second node shows:
```erlang
{[ejabberd2@localhost,ejabberd@localhost],
 [ejabberd@localhost],
 [mqtt_session,push_session,last_activity,pubsub_index,
  vcard_search,oauth_client,archive_prefs,motd_users,mqtt_sub,
  reg_users_counter,mqtt_pub,sr_user,schema,roster_version,
  pubsub_orphan,session,pubsub_last_item,offline_msg,
  archive_msg,route,private_storage,motd,sr_group,oauth_token,
  bytestream,muc_room|...]}
```

Conclusions:
- each mnesia table can have a different master
- only a slave node knows what is its master. A master doesn't know who are its slaves
- `join_cluster` doesn't setup a master by default
- using `set_master` you can define one or several nodes from where the tables will be downloaded in case of network partition. In fact you can have cyclic master/slaves

**Comment by joudinet** [comment]
Thanks for your explanation! I haven't realized the purpose of master/slave in mnesia until now. I guess I should use set_master with parsimony. I'm not sure if it makes sense to say every node is the slave of all other nodes in its cluster? I don't have the guarantee that any master node will be available when a network partition is detected. If I `set_master` to all nodes in the cluster, but none are available on a network partition, do I still get the mnesia event in ejabberd's error log about the inconsistent_database?

I modified the way I treat a network partition. Now, I delete the local database and restart a fresh ejabberd node. Since I have done this modification, I'm unable to reproduce the bug I described here. I'm closing this bug and if it shows up again, I'll try to play with the `set_master` command. Thanks again for your help.
